pkg/chunk/aws_storage_client.go

```diff
@@ -331,28 +340,232 @@ func (a dynamoDBRequestAdapter) HasNextPage() bool {
 	}
-	return buf, nil
+	return chunk, nil
 }
 
-func (a awsStorageClient) PutChunk(ctx context.Context, key string, buf []byte) error {
+func (a awsStorageClient) getDynamoDBChunks(ctx context.Context, chunks []Chunk) ([]Chunk, error) {
+	outstanding := dynamoDBReadRequest{}
+	chunksByKey := map[string]Chunk{}
+	for _, chunk := range chunks {
+		key := chunk.externalKey()
+		chunksByKey[key] = chunk
+		tableName := a.chunkTableFor(chunk.From)
+		outstanding.Add(tableName, key, nil)
+	}
+
+	result := []Chunk{}
+	unprocessed := dynamoDBReadRequest{}
+	backoff, numRetries := minBackoff, 0
+	for outstanding.Len()+unprocessed.Len() > 0 {
+		requests := dynamoDBReadRequest{}
+		requests.takeReqs(unprocessed, dynamoDBMaxReadBatchSize)
+		requests.takeReqs(outstanding, dynamoDBMaxReadBatchSize)
+
+		var response *dynamodb.BatchGetItemOutput
+		err := instrument.TimeRequestHistogram(ctx, "DynamoDB.BatchGetItemPages", dynamoRequestDuration, func(ctx context.Context) error {
+			var err error
+			response, err = a.DynamoDB.BatchGetItemWithContext(ctx, &dynamodb.BatchGetItemInput{
+				RequestItems:           outstanding,
+				ReturnConsumedCapacity: aws.String(dynamodb.ReturnConsumedCapacityTotal),
+			})
+			return err
+		})
+
+		for _, cc := range response.ConsumedCapacity {
+			dynamoConsumedCapacity.WithLabelValues("DynamoDB.BatchGetItemPages").
+				Add(float64(*cc.CapacityUnits))
+		}
+
+		if err != nil {
+			for tableName := range outstanding {
+				recordDynamoError(tableName, err)
+			}
+		}
+
+		for _, items := range response.Responses {
+			for _, item := range items {
+				key, ok := item[hashKey]
+				if !ok || key == nil || key.S == nil {
+					return result, fmt.Errorf("Got response from DynamoDB with no hash key: %+v", item)
+				}
+
+				chunk, ok := chunksByKey[*key.S]
+				if !ok {
+					return result, fmt.Errorf("Got response from DynamoDB with for chunk I didn't ask for: %s", *key.S)
+				}
+
+				buf, ok := item[valueKey]
+				if !ok || buf == nil || buf.B == nil {
+					return result, fmt.Errorf("Got response from DynamoDB with no value: %+v", item)
+				}
+
+				if err := chunk.decode(buf.B); err != nil {
+					return result, err
+				}
+
+				result = append(result, chunk)
+			}
+		}
+
+		// If there are unprocessed items, backoff and retry those items.
+		if unprocessedKeys := response.UnprocessedKeys; unprocessedKeys != nil && dynamoDBReadRequest(unprocessedKeys).Len() > 0 {
+			unprocessed.takeReqs(unprocessedKeys, -1)
+			time.Sleep(backoff)
+			backoff = nextBackoff(backoff)
+			continue
+		}
+
+		// If we get provisionedThroughputExceededException, then no items were processed,
+		// so back off and retry all.
+		if awsErr, ok := err.(awserr.Error); ok && awsErr.Code() == provisionedThroughputExceededException {
+			unprocessed.takeReqs(requests, -1)
+			time.Sleep(backoff)
+			backoff = nextBackoff(backoff)
+			numRetries++
+			continue
+		}
+
+		// All other errors are fatal.
+		if err != nil {
+			return result, err
+		}
```

Comment:
Wouldn't it make more sense to push these to immediately after we get the error, just after we do the "record dynamodb error" bit? As it is, it's disconnected from that logic, and it's not clear whether it matters that this happens _after_ turning the responses into chunks.